activerecord/lib/active_record/relation/finder_methods.rb

```diff
@@ -407,7 +407,7 @@ def apply_join_dependency(relation, join_dependency)
       relation = relation.except(:includes, :eager_load, :preload)
       relation = relation.joins join_dependency
 
-      if using_limitable_reflections?(join_dependency.reflections)
+      if using_limitable_reflections?(join_dependency.reflections) || relation.having_clause.any?
```

Comment:
The presence of a HAVING clause doesn't make an otherwise-not-limitable reflection into one that can safely support a limit.

This was already the preferred path: if the existing condition has decided it won't work, we'd need a very well-considered argument to introduce an exception.